### PR TITLE
Button/Box

### PR DESCRIPTION
**Pull request by shaobinspire**

### CODE REVIEW

File: Applications/Spire/Include/Spire/Ui/Box.hpp [hunk 1]
```diff
@@ -0,0 +1,86 @@
+#ifndef SPIRE_BOX_HPP
+#define SPIRE_BOX_HPP
+#include <boost/optional.hpp>
+#include <boost/variant.hpp>
+#include <QWidget>
+#include "Spire/Ui/Ui.hpp"
+
+namespace Spire {
+
+  //! Represents a box container with a spcific styling.
```
Comment (spireken) [inline Applications/Spire/Include/Spire/Ui/Box.hpp:10]: Typo: specific.

File: Applications/Spire/Include/Spire/Ui/TextBox.hpp [hunk 2]
```diff
@@ -68,16 +101,20 @@ namespace Spire {
       void focusOutEvent(QFocusEvent* event) override;
       void keyPressEvent(QKeyEvent* event) override;
       void resizeEvent(QResizeEvent* event) override;
+      bool event(QEvent* event) override;
       QSize sizeHint() const override;
 
     private:
       mutable CurrentSignal m_current_signal;
       mutable SubmitSignal m_submit_signal;
+      bool m_is_hover;
```
Comment (spireken) [inline Applications/Spire/Include/Spire/Ui/TextBox.hpp:110]: m_is_hovered

File: Applications/Spire/Include/Spire/Ui/Box.hpp [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+      virtual void process_hover_style(const Style& hover_style);
+      virtual void process_focus_style(const Style& focus_style);
+      virtual void process_disabled_style(const Style& disabled_style);
+
+    private:
+      Style m_style;
+      Style m_hover_style;
+      Style m_focus_style;
+      Style m_disabled_style;
+      bool m_is_hover;
```
Comment (spireken) [inline Applications/Spire/Include/Spire/Ui/Box.hpp:77]: m_is_hovered

**Review by spireken — CHANGES_REQUESTED** [review]

**Comment by spireken** [comment]
An issue I found; the button is supposed to trigger the ClickSignal on a space bar press and release. Currently, it triggers with just a space bar press.

**Comment by shaobinspire** [comment]
> the button is supposed to trigger the ClickSignal on a space bar press and release.

I implemented it according to the implementation of the FlatButton class. No matter In the current flatbutton specs or the button specs, this is not mentioned. If the button triggers the ClickSignal on a space bar press and release, it means that one click of the space bar will trigger the signal twice. I think it’s abnormal.

**Comment by spireken** [comment]
It's a subtle detail, but in the flow for the Button specs the flow makes a distinction between an enter 'press' event and a space 'click' event. A 'click' for a key event is a full press/release cycle.

I believe this was based off of certain buttons that made a distinction between the enter/return keys and the space bar, but you would need to ask the designers for the full details.

**Pull request closed without merging**